designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+# UX: S3 explorer
+
+* **Type**: UX Design
+* **Status**: Proposal
+
+## Overview
+
+This document proposes a user experience for interacting with AWS S3 from the
+AWS Toolkit ("the Toolkit") for Visual Studio Code ("VSCode").
+
+Some of the features listed below are vague; they are implemented using VSCode
+or platform conventions analogous to the AWS web console.
+
+## Experience
+
+Each region of the Toolkit explorer tree includes an S3 node (see mockup
+below). The toolkit pulls a list of names and properties of S3 objects in the
+currently active profile and presents this data as a tree in the toolkit
+explorer. This is typical of existing toolkit features such as "Schemas" and
+"CDK".
+
+- When a region is expanded, make an eager request to get the count of S3
+  buckets for that region.
+- Show the count of buckets next to the S3 node.
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
+- When the current active profile changes, the tree is not refreshed, but is
```

Comment:
Why not? What if the user changes accounts, or changes to a user that doesn't have bucket permissions?

Reply:
In general I am advocating that we avoid clobbering the current state just because the user may be exploring some other account.

We should think about copying objects across accounts too. Initially this might look similar to how we handle regions: yet another parent-node in the Toolkit explorer. There are ways we can avoid cluttering the UX for the common case.

Reply:
I find this interesting also. You might change profiles to a different account id.

Reply:
I think clobbering is the right way to go when we don't have knowledge of what the other account has access to. Swapping profiles is a direct user action, so I'd feel that a user would be expecting for all of these fields to change.